lightning/src/ln/channelmanager.rs

```diff
@@ -4740,7 +4896,14 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
 		}
 		let mut channel_state_lock = self.channel_state.lock().unwrap();
 		let channel_state = &mut *channel_state_lock;
-		match channel_state.by_id.entry(funding_msg.channel_id) {
+		let per_peer_state = self.per_peer_state.read().unwrap();
+		if let None = per_peer_state.get(counterparty_node_id) {
+			return Err(MsgHandleErrInternal::send_err_msg_no_close(format!("The peer with node_id {}, was dropped before the signed funding tx was sent to the peer", counterparty_node_id), funding_msg.channel_id))
```

Comment:
Wasn't 100% sure how to handle this. Could potentially happen if this was the only channel and the peer disconnects (executing `peer_disconnected`) between removing the channel above and retaking the `per_peer_state` here IIUC. Taking the `read` lock throughout this function doesn't help either, as `no_channels_remain` will be `true` in `peer_disconnected` if the case I explained above occurs, effectively making that function remove the `peer` from the `per_peer_state`.

Let me know if you have a better suggestion of how we'd want to handle this.

Reply:
> We definitely need to check if channels remain and remove the peer state itself under the same lock in `peer_disconnected`. I don't think we should over-optimize for the case where a channel remains too much there, either - hopefully our channel peers aren't disconnecting + reconnecting in a tight loop :)

Ah yeah of course, my bad :)! Updated `peer_disconnected` with [ae4ec21](https://github.com/lightningdevkit/rust-lightning/pull/1507/commits/ae4ec219b042ba756f3ceb38abd1f0c755431a76).
That will also mean that we will only need to hold the `read` lock throughout this function, and can drop the inner `PeerState` during `chain_monitor.watch_channel` in `internal_funding_created`. Updated that with [bf72c90](https://github.com/lightningdevkit/rust-lightning/pull/1507/commits/bf72c903b1053a58e1949d39ce7575e521a402f4).